### PR TITLE
Emit a warning rather than crash when a message is too big for sqlite

### PR DESCRIPTION
**Pull request by wjwwood**
Related to issue https://github.com/ros2/rosbag2/issues/915

This pull request changes the behavior of the sqlite storage plugin to emit a warning rather than throw an exception when a message is written that is too large for sqlite.

### CODE REVIEW

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp [hunk 1]
```diff
@@ -248,7 +248,21 @@ void SqliteStorage::write_locked(
             "' has not been created yet! Call 'create_topic' first.");
   }
 
-  write_statement_->bind(message->time_stamp, topic_entry->second, message->serialized_data);
+  try {
+    write_statement_->bind(message->time_stamp, topic_entry->second, message->serialized_data);
+  } catch (const SqliteException & exc) {
+    if (SQLITE_TOOBIG == exc.get_sqlite_return_code()) {
+      ROSBAG2_STORAGE_DEFAULT_PLUGINS_LOG_WARN_STREAM(
```
Comment (wjwwood) [inline rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:255]: I thought about using one of the throttle or once type macros, but I hesitated for this since it's simpler. But I'm open to suggestions on what to do and, if throttle, how often to throttle it.

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp [hunk 2]
```diff
@@ -156,22 +165,58 @@ class StorageTestFixture : public TemporaryDirectoryFixture
...
+
+      amount_to_write = remaining_buffer;
+      if (message.size() < amount_to_write) {
+        amount_to_write = message.size();
+      }
+      std::memcpy(buffer + amount_written, message.data(), amount_to_write);
+      amount_written += amount_to_write;
+
+      return amount_written;
+    }
```
Comment (wjwwood) [inline rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp:219]: So, I could have just removed the snprintf section, since I _think_ my `memcpy` one is equivalent, but I just wanted to limit my side effects with this change. Let me know what you guys think.

**Comment by wjwwood** [comment]
This is what the error message looks like in practice:

```
[WARN] [1637792715.870418893] [rosbag2_storage]: Message on topic '/too_big_message' of size '1000000010' bytes failed to write because it exceeds the maximum size sqlite can store: Error when binding SQL parameter 3. SQLite error (18): string or blob too big
```

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp (hunk 2 above)
Comment (gbiggs) [reply rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp:219]: It looks equivalent to me. I'd be fine with removing the `snprintf` version to keep the code simpler.

**Review by gbiggs — APPROVED** [review]

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp (hunk 1 above)
Comment (danthony06) [reply rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:255]: I agree with @gbiggs. And practically, it seems pretty unlikely that you'll have messages exceeding the maximum SQLite blob size that are also coming at a high rate, because you would need an unusually fast network connection for that to happen.

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp [hunk 3]
```diff
@@ -248,7 +248,21 @@ void SqliteStorage::write_locked(
...
+        "Message on topic '" << message->topic_name << "' of size '"
+        << message->serialized_data->buffer_length
+        << "' bytes failed to write because it exceeds the maximum size sqlite can store: "
```
Comment (danthony06) [inline rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:258]: Could we also include the SQLITE_MAX_LENGTH (https://www.sqlite.org/limits.html) value in this error message, so that the users can know what it's set at?

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/test_sqlite_storage.cpp [hunk 4]
```diff
@@ -373,3 +373,23 @@ TEST_F(StorageTestFixture, throws_on_invalid_pragma_in_config_file) {
...
+
+  // Get the sqlite string/blob limit.
+  size_t sqlite_limit = sqlite3_limit(
+    writable_storage->get_sqlite_database_wrapper().get_database(),
+    SQLITE_LIMIT_LENGTH,
+    -1);
+
+  // This should produce a warning, but not an exception.
+  std::string msg(sqlite_limit + 1, '\0');
+  this->write_messages_to_sqlite(
```
Comment (emersonknapp) [inline rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/test_sqlite_storage.cpp:391]: It doesn't actually affect the functionality of the test, but I find it helps readability to use `EXPECT_NO_THROW({code_block...})` - then you don't have to use a comment.

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp (hunk 2 above)
Comment (emersonknapp) [reply rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp:219]: Also fine with changing to purely `memcpy` to keep things consistent.

**Review by emersonknapp — APPROVED** [review]
This LGTM

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp (hunk 3 above)
Comment (wjwwood) [reply rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:258]: So, this value doesn't appear to be in the headers from sqlite (it's a compiler definition, but it isn't stored in a header AFAIK). I can get it dynamically by "changing" the limit too `-1` and it will return the current value without modifying it, but this requires an sqlite database and I'm not sure I have one in scope here. I'll try.

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp (hunk 2 above)
Comment (wjwwood) [reply rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp:219]: Ok, will do.

File: rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp (hunk 3 above)
Comment (danthony06) [reply rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp:258]: Ah, that's too bad. I don't think it's really critical, just a nice thing to know when it happens without needing to go to the documentation.

Comment (wjwwood) [reply rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp]: I got it, I forgot we were in a method of the class that holds the db.

**Comment by wjwwood** [comment]
I had to refactor more of the test fixture's methods to avoid null terminating characters now that we're using `memcpy` and not `snprintf`.

**Review by gbiggs — APPROVED** [review]

**Review by emersonknapp — APPROVED** [review]

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/701c5a90a6259e55494a5ecc055e44e3/raw/fe5656f5fc29f36e5f5c17b3152efece41fee6dd/ros2.repos
BUILD args: --packages-up-to rosbag2_storage_default_plugins rosbag2_tests rosbag2
TEST args: --packages-select rosbag2_storage_default_plugins rosbag2_tests rosbag2
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/9437
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15711/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10393/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13376/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15953/)

**Comment by wjwwood** [comment]
I'm looking into the test failures on macOS and Windows.

**Comment by wjwwood** [comment]
Re-running CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15717/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10399/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13382/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15961/)

**Comment by wjwwood** [comment]
@ros-pull-request-builder test this please

**Comment by wjwwood** [comment]
So, either because of how Homebrew builds sqlite or because of a newer version of sqlite, the limit for a blob on my macOS machine and our buildfarm is just over 17GB. Which is causing a new failure in how we use the API. I'm gonna try one thing, but if that doesn't work, I might have to skip this test on machines where the limit is more than say 2GB.

**Comment by wjwwood** [comment]
Ok, I found a solution in 51cb11c (and improved in 3f433b6), where I artificially lower the limit. This should work on all the machines.

New CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15719/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10401/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13384/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15963/)

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp [hunk 5]
```diff
@@ -156,22 +161,45 @@ class StorageTestFixture : public TemporaryDirectoryFixture
...
-      message.c_str());
+    size_t remaining_buffer = buffer_capacity;
+    size_t amount_written = 0;
+    if (remaining_buffer == 0) {
+      return amount_written;
+    }
+
+    size_t amount_to_write = remaining_buffer;
+    std::string preamble = get_preamble();
+    if (preamble.size() < amount_to_write) {
```
Comment (hidmic) [inline rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp:185]: @wjwwood nit: would it make sense to `assert(buffer_capacity >= preamble.size() + message.size())`?

File: rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp [hunk 6]
```diff
@@ -73,48 +73,53 @@ class StorageTestFixture : public TemporaryDirectoryFixture
       });
 
     serialized_data->buffer_length = message_size;
-    int written_size = write_data_to_serialized_string_message(
+    size_t written_size = write_data_to_serialized_string_message(
```
Comment (hidmic) [inline rosbag2_storage_default_plugins/test/rosbag2_storage_default_plugins/sqlite/storage_test_fixture.hpp:76]: @wjwwood would 
```suggestion
    (void)write_data_to_serialized_string_message(
```
have the same effect?

**Review by hidmic — APPROVED** [review]
Minor comments, all of them.

**Comment by wjwwood** [comment]
@hidmic I addressed your feedback and fixed one more warning on windows, so if you could quickly re-review those I'd appreciate it.

CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15894/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10569/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13558/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16201/)

**Comment by wjwwood** [comment]
CI again after fixing a new bug:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15905/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10580/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13569/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16225/)

**Comment by wjwwood** [comment]
The `ament_uncrustify` lint will fail until rolling is synced, which is a known issue. Locally it passes.

**Review by sloretz — APPROVED** [review]
Approving just the two new commits (https://github.com/ros2/rosbag2/pull/919/commits/52f96e2ec18421be0d842151dc76f8515172ca28 , and https://github.com/ros2/rosbag2/pull/919/commits/ebdce6700324d82d9f65e4a3c7a1d719780985ad ) , I haven't reviewed the rest
